src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/DoNotUseCompilationGetSemanticModelAnalyzer.cs

```diff
@@ -0,0 +1,73 @@
+
+        public static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticIds.DoNotUseCompilationGetSemanticModelRuleId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            DiagnosticHelpers.DefaultDiagnosticSeverity,
+            isEnabledByDefault: DiagnosticHelpers.EnabledByDefaultIfNotBuildingVSIX,
+            description: s_localizableDescription,
+            customTags: WellKnownDiagnosticTags.Telemetry);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(csac =>
+            {
+                var wellKnownTypeProvider = WellKnownTypeProvider.GetOrCreate(csac.Compilation);
+
+                if (!wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisDiagnosticsDiagnosticAnalyzer, out var diagnosticAnalyzerType) ||
+                    !wellKnownTypeProvider.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisCompilation, out var compilationType))
+                {
+                    return;
+                }
+
+                var csharpCompilation = wellKnownTypeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisCSharpCSharpCompilation);
+                var visualBasicCompilation = wellKnownTypeProvider.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisVisualBasicVisualBasicCompilation);
+
+                csac.RegisterOperationBlockStartAction(obsac =>
+                {
+                    if (obsac.OwningSymbol is IMethodSymbol methodSymbol &&
+                        methodSymbol.ContainingType.Inherits(diagnosticAnalyzerType))
+                    {
+                        obsac.RegisterOperationAction(oac =>
+                        {
+                            var invocation = (IInvocationOperation)oac.Operation;
+
+                            if (invocation.TargetMethod.Name.Equals("GetSemanticModel", StringComparison.Ordinal) &&
+                                (
+                                    invocation.TargetMethod.ContainingType.Equals(compilationType) ||
+                                    invocation.TargetMethod.ContainingType.Equals(csharpCompilation) ||
+                                    invocation.TargetMethod.ContainingType.Equals(visualBasicCompilation)
```

Comment:
Do you have any preference between these 3 OR conditions and a local method?

Reply:
I think this is fine.